### PR TITLE
Enabling helper methods to merge a template with an override.

### PR DESCRIPTION
**Pull request by n3wscott**

### CODE REVIEW

File: pkg/manifest/templates.go [hunk 1]
```diff
@@ -102,15 +107,25 @@ func ParseTemplates(path string, images map[string]string, cfg map[string]interf
 }
 
 // ExecuteLocalYAML will look in the callers filesystem and process the
-// templates found in files named "*.yaml" and return the f.
+// templates found in files named "*.yaml".
+// Depreciated: use ExecuteYAML(images, cfg)
```
Comment (vaikas) [inline pkg/manifest/templates.go:111]: nit, deprecated.

File: pkg/manifest/templates.go [hunk 2]
```diff
@@ -137,7 +152,73 @@ func removeComments(in string) string {
 	return regex.ReplaceAllString(in, "")
 }
 
-// OutputYAML writes out each file contents  to out after removing comments and
+func MergeYAML(files map[string]string, overlays map[string]string) ([]unstructured.Unstructured, error) {
```
Comment (vaikas) [inline pkg/manifest/templates.go:155]: I think this could use a comment or two on what it does.

File: pkg/manifest/templates_test.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	results, err := manifest.MergeYAML(files, overrides)
+	if err != nil {
+		panic(err)
+	}
+
+	manifest.OutputUnstructuredAsYAML(os.Stdout, results)
```
Comment (vaikas) [inline pkg/manifest/templates_test.go:134]: Could we add a test that we know we're going to need for Broker. Namely, writing the metadata.annotations field?

Comment (n3wscott) [reply pkg/manifest/templates_test.go:134]: added.

File: pkg/manifest/templates.go (hunk 2 above)
Comment (n3wscott) [reply pkg/manifest/templates.go:159]: added a comment

File: pkg/manifest/templates.go (hunk 1 above)
Comment (n3wscott) [reply pkg/manifest/templates.go]: idk how to spell, sorry

File: pkg/manifest/templates.go [hunk 4]
```diff
@@ -102,15 +107,25 @@ func ParseTemplates(path string, images map[string]string, cfg map[string]interf
...
 func ExecuteLocalYAML(images map[string]string, cfg map[string]interface{}) (map[string]string, error) {
-	pwd, _ := os.Getwd()
-	log.Println("PWD: ", pwd)
 	_, filename, _, _ := runtime.Caller(1)
 
 	return ExecuteTemplates(path.Dir(filename), "yaml", images, cfg)
 }
 
+// ExecuteYAML will look in the callers filesystem and process the
+// templates found in files named "./<f/r/o/m>/*.yaml". Relative to the caller.
```
Comment (vaikas) [inline pkg/manifest/templates.go:119]: Why do we want it to be relative to the caller? Since these are mainly consumed by conformance tests where pieces come that are not related to caller, seems odd?

Comment (n3wscott) [reply pkg/manifest/templates.go:119]: fair, let me think about it

Comment (slinkydeveloper) [reply pkg/manifest/templates.go:119]: To me it seems reasonable to have it relative to the caller. Maybe we need another one where from is absolute (and not a vararg maybe?)

File: pkg/manifest/templates.go [hunk 5]
```diff
@@ -102,15 +107,25 @@ func ParseTemplates(path string, images map[string]string, cfg map[string]interf
 }
 
 // ExecuteLocalYAML will look in the callers filesystem and process the
-// templates found in files named "*.yaml" and return the f.
+// templates found in files named "*.yaml".
+// Deprecated: use ExecuteYAML(images, cfg)
 func ExecuteLocalYAML(images map[string]string, cfg map[string]interface{}) (map[string]string, error) {
```
Comment (slinkydeveloper) [inline pkg/manifest/templates.go:112]: Why this is deprecated? It seems to me an alternative of `ExecuteYAML`

File: pkg/manifest/templates_test.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	// Output:
```
Comment (slinkydeveloper) [inline pkg/manifest/templates_test.go:177]: Is it expected that there isn't any `Foo` there?

**Review by slinkydeveloper — COMMENTED** [review]
Is there any reason for the yaml to merge to be the "full" spec + the changes? Wouldn't it be better to just specify the difference, eg:

base:

```yaml
apiVersion: example.knative.dev/v1
kind: AAA
metadata:
  name: aaa-{{ .name }}
  namespace: {{ .namespace }}
spec:
  aaa: "{{ .aaaMsg }}"
```

Override:

```
metadata:
  annotations:
    some-custom-annotation: for all AAAs
```

Identifying what has to be overriden can be done of name basis or some key that identifies the template "id".

Also, given you use the GVK to identify which templates to merge, what happens if you need to have several yamls for the same GVK but different resource (e.g. different name/namespace)?
